common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -386,6 +444,46 @@ protected final boolean runAllTasksFrom(Queue<Runnable> taskQueue) {
         }
     }
 
+    /**
+     * What ever tasks are present in {@code taskQueue} when this method is invoked will be {@link Runnable#run()}.
+     * @param taskQueue the task queue to drain.
+     * @return {@code true} if at least {@link Runnable#run()} was called.
+     */
+    private boolean runExistingTasksFrom(Queue<Runnable> taskQueue) {
+        return taskQueue.offer(BOOKEND_TASK) ? runExistingTasksUntilBookend(taskQueue)
+                                             : runExistingTasksUntilMaxTasks(taskQueue);
+    }
+
+    private boolean runExistingTasksUntilBookend(Queue<Runnable> taskQueue) {
+        Runnable task = pollTaskFrom(taskQueue);
+        if (task == BOOKEND_TASK) {
+            return false;
+        }
+        for (;;) {
+            safeExecute(task);
+            task = pollTaskFrom(taskQueue);
+            if (task == BOOKEND_TASK) {
+                return true;
+            }
+        }
+    }
+
+    private boolean runExistingTasksUntilMaxTasks(Queue<Runnable> taskQueue) {
+        Runnable task = pollTaskFrom(taskQueue);
+        if (task == null) {
+            return false;
+        }
+        int i = 0;
+        do {
+            safeExecute(task);
+            task = pollTaskFrom(taskQueue);
+            if (task == null) {
```

Comment:
Should we also respect `BOOKEND_TASK ` here, or do we not expect to have one in here as we failed to add it before calling this and only will ever have one in the queue ?

Reply:
Right. `BOOKEND_TASK` shouldn't be encountered here as you indicated.

Reply:
maybe we should add an `assert` ?

Reply:
We can if you prefer however if we don't nothing negative will happen.

Reply:
Can you just add a comment for it ?

Reply:
done